apps/tasks/views.py

```diff
@@ -1,177 +1,158 @@
 
 
-class TaskViewSet(ModelViewSet):
-    serializer_class = TaskListSerializer
+class TaskViewSet(ViewSet,
+                  mixins.ListModelMixin,
+                  mixins.RetrieveModelMixin,
+                  mixins.CreateModelMixin,
+                  mixins.DestroyModelMixin,
+                  GenericViewSet):
     queryset = Task.objects.all()
+    serializer_class = TaskSerializer
     permission_classes = (IsAuthenticated,)
-    filter_backends = [SearchFilter]
-    search_fields = ['title',]
-
-    def get_serializer_class(self):
-
-        if self.action in ['update', 'partial_update']:
-            return TaskUpdateSerializer
-
-        if self.action in ['mytask', 'completed', 'complete']:
-            return Serializer
-
-        if self.action in ['assign']:
-            return TaskAssignSerializer
+    search_fields = ['title']
+    filterset_fields = ('status', 'assigned_to')
 
-        if self.action in ['search']:
-            return SearchTaskSerializer
-
-        if self.action in ['list']:
-            return TaskListSerializer
+    def get_queryset(self):
+        qs = super().get_queryset()
+        if self.action in ['list', 'top20']:
+            return qs.annotate(total_duration=(Sum('timelog_task__duration'))).order_by('-id')
+        if self.action in ['start', 'stop']:
+            return Timer.objects.all()
+        return qs
 
+    def get_serializer_class(self):
+        if self.action in ['list', 'top20']:
+            return TaskWithDurationSerializer
+        if self.action in ['start', 'stop']:
+            return None
         return TaskSerializer
 
     def perform_create(self, serializer):
         serializer.save(created_by=self.request.user)
 
-    @action(detail=True, methods=['POST'])
-    def complete(self, request, pk=None):
+    @method_decorator(cache_page(CACHE_TTL))
+    @action(detail=False)
+    def top20(self, response, *args, **kwargs):
+        this_month = datetime.now().replace(day=1).date()
+        tasks = (self.get_queryset()
+                 .filter(timelog_task__started_at__gte=this_month)
+                 .filter(total_duration__isnull=False)
+                 .order_by('-total_duration')[:20]
+                 )
+        tasks_data = self.get_serializer(tasks, many=True).data
+        return Response(tasks_data)
+
+    @action(detail=True, methods=['POST'], serializer_class=Serializer)
+    def complete(self, request, *args, **kwargs):
         instance = self.get_object()
         instance.status = 'completed'
         instance.save()
-        send_mail(
-            receivers=instance.assigned_to.email,
-            subject="You have ben assigned a new task!",
-            message=f"You have ben assigned a new task!\n The new task is \"{instance.title}\".")
-        return Response(data={'details': 'Task completed'})
+        serializer = self.get_serializer(instance=instance)
+        return Response(serializer.data)
 
     @action(detail=False, serializer_class=Serializer)
```

Comment:
In cerinte scrie ca trebuie endpoint aparte? Acest lucru se poate de inlocuit cu filtrele